modules/sor/sorV2/lib/poolsV3/buffer/bufferPool.ts

```diff
@@ -0,0 +1,115 @@
+import { Address, Hex } from 'viem';
+import { MAX_UINT256, SwapKind, Token, TokenAmount } from '@balancer/sdk';
+import { BufferState, Vault } from '@balancer-labs/balancer-maths';
+
+import { BasePoolV3 } from '../../poolsV2/basePool';
+import { BasePoolToken } from '../../poolsV2/basePoolToken';
+import { Erc4626PoolToken } from '../../poolsV2/erc4626PoolToken';
+
+export class BufferPool implements BasePoolV3 {
+    public readonly chainId: number;
+    public readonly id: Hex;
+    public readonly address: string;
+    public readonly poolType = 'Buffer';
+    public readonly swapFee = 0n;
+    public readonly rate: bigint;
+    public readonly tokens: BasePoolToken[];
+
+    private readonly tokenMap: Map<string, BasePoolToken>;
+
+    static fromErc4626Token(erc4626Token: Erc4626PoolToken): BufferPool {
+        const mainToken = new BasePoolToken(erc4626Token.token, MAX_UINT256, 0);
+        const underlyingToken = new BasePoolToken(
+            new Token(
+                erc4626Token.token.chainId,
+                erc4626Token.underlyingTokenAddress as Address,
+                erc4626Token.token.decimals,
+            ),
+            MAX_UINT256,
+            1,
+        );
+        return new BufferPool(
+            erc4626Token.token.address,
+            erc4626Token.token.address,
+            erc4626Token.token.chainId,
+            erc4626Token.rate,
+            [mainToken, underlyingToken],
+        );
+    }
+
+    constructor(id: Hex, address: string, chainId: number, rate: bigint, tokens: BasePoolToken[]) {
+        this.chainId = chainId;
+        this.id = id;
+        this.address = address;
+        this.rate = rate;
+        this.tokens = tokens;
+        this.tokenMap = new Map(tokens.map((token) => [token.token.address, token]));
+    }
+
+    public getNormalizedLiquidity(tokenIn: Token, tokenOut: Token): bigint {
+        return MAX_UINT256;
+    }
+
+    public getLimitAmountSwap(tokenIn: Token, tokenOut: Token, swapKind: SwapKind): bigint {
+        return MAX_UINT256;
+    }
+
+    public swapGivenIn(tokenIn: Token, tokenOut: Token, swapAmount: TokenAmount): TokenAmount {
+        const { tIn, tOut } = this.getRequiredTokenPair(tokenIn, tokenOut);
+
+        const poolState = this.getPoolState();
+        const vault = new Vault();
+
+        const calculatedAmount = vault.swap(
+            {
+                amountRaw: swapAmount.amount,
+                tokenIn: tIn.token.address,
+                tokenOut: tOut.token.address,
+                swapKind: SwapKind.GivenIn,
+            },
+            poolState,
+        );
+        return TokenAmount.fromRawAmount(tOut.token, calculatedAmount);
+    }
+
+    public swapGivenOut(tokenIn: Token, tokenOut: Token, swapAmount: TokenAmount): TokenAmount {
+        const { tIn, tOut } = this.getRequiredTokenPair(tokenIn, tokenOut);
+
+        const poolState = this.getPoolState();
```

Comment:
`poolState` and `vault` never change after initial instantiation do they? So we could maybe just call them once in constructor and reuse?